src/main/java/nextstep/security/web/csrf/CsrfFilter.java

```diff
@@ -0,0 +1,74 @@
+package nextstep.security.web.csrf;
+
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import nextstep.security.access.MvcRequestMatcher;
+import nextstep.security.access.RequestMatcher;
+import org.springframework.util.Assert;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+public class CsrfFilter extends OncePerRequestFilter {
+    public static final RequestMatcher DEFAULT_CSRF_MATCHER = new DefaultRequiresCsrfMatcher();
+
+    private RequestMatcher requireCsrfProtectionMatcher = DEFAULT_CSRF_MATCHER;
+    private final AccessDeniedHandler accessDeniedHandler = new AccessDeniedHandler();
+    private final CsrfTokenRepository tokenRepository;
+
+    public CsrfFilter(CsrfTokenRepository tokenRepository) {
+        Assert.notNull(tokenRepository, "tokenRepository cannot be null");
+        this.tokenRepository = tokenRepository;
+    }
+    public void setRequireCsrfProtectionMatcher(RequestMatcher requireCsrfProtectionMatcher) {
+        Assert.notNull(requireCsrfProtectionMatcher, "requireCsrfProtectionMatcher cannot be null");
+        this.requireCsrfProtectionMatcher = requireCsrfProtectionMatcher;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+            throws ServletException, IOException {
+        CsrfToken csrfToken = this.tokenRepository.loadToken(request);
+        boolean missingToken = (csrfToken == null);
+        if (missingToken) {
+            csrfToken = this.tokenRepository.generateToken(request);
+            this.tokenRepository.saveToken(csrfToken, request, response);
+        }
+        request.setAttribute(CsrfToken.class.getName(), csrfToken);
+        if (!this.requireCsrfProtectionMatcher.matches(request)) {
+            filterChain.doFilter(request, response);
+            return;
+        }
+        String actualToken = request.getHeader(csrfToken.getHeaderName());
+        if (actualToken == null) {
+            actualToken = request.getParameter(csrfToken.getParameterName());
+        }
+        if (!Objects.equals(csrfToken.getToken(), actualToken)) {
+            this.accessDeniedHandler.handle(request, response);
+            return;
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private static final class DefaultRequiresCsrfMatcher implements RequestMatcher {
+
+        private final HashSet<String> allowedMethods = new HashSet<>(Arrays.asList("GET", "HEAD", "TRACE", "OPTIONS"));
```

Comment:
소소하지만 다음과 같이 해볼 수 있겠네요 😄 
```suggestion
        private final HashSet<String> allowedMethods =     new HashSet<>(Arrays.asList(HttpMethod.GET.name(), HttpMethod.HEAD.name(),  HttpMethod.TRACE.name(), HttpMethod.OPTIONS.name()));
```